### DISCUSSION TITLE
Alternative endings for alphaTex

### DISCUSSION THREAD

**Discussion by jonaro00** [CoderLine/alphaTab#671] [Ideas]
Hi,
Thanks for making this project. I'm having very fun playing with it.

The only feature (yet) I wish alphaTex had support for is Alternative endings to repeats. Would appreciate if it was added.
For reference: http://lilypond.org/doc/v2.23/Documentation/notation/long-repeats#alternative-endings

Perhaps the syntax could be:
```
\ro 3.3.1 | \ae 1 \ae 3 3.3.2*2 | \ae 2 \ae 4 \rc 4 3.3.4*4 |
```
but I haven't spend much time checking how the parser works and what it can realistically handle.

What do you think?

P.S. Is there a conventional file name extension for alphaTex files?

**Comment by Danielku15** [comment]
Indeed alternative endings are a missing feature until now in alphaTex. To indicate better the relation on repeats I would maybe rather go for a metadata tag like `\re` (repeat endings). Instead of requiring `\re` to be put in front of each ending, I would maybe go for a syntax like:

```
\ro 3.3.1 | \re 1 3.2.2*2 | \re (2 4) \rc 4 3.3.4*4 |
```

Looking at the overall alphaTex syntax, in many effects we use parenthesis to group multiple values. The parenthesis can be skipped if there is only 1 value. The parser is implemented a recursive decent parser and extending it should be usually quite simple, you can give it a try if you're interested into adding this feature. With a bit of reading you should be able to understand quite quickly how the logic works. 

Example on how beats handle the parsing of either single or multiple values in parenthesis: 
https://github.com/CoderLine/alphaTab/blob/develop/src/importer/AlphaTexImporter.ts#L1080

Place where the existing repeat metadata tags are handled:
https://github.com/CoderLine/alphaTab/blob/develop/src/importer/AlphaTexImporter.ts#L1775

The rest of alphaTab should already take care that the repeats are handled correctly. Of course if somebody would write messy repeats, alphaTab would behave like this too. 

There is no file extension for alphaTex because I never envisioned it as format to be written into files, but rather to store it in databases or providing it easily through APIs. If I would need to think of one, I guess it would be maybe `.atx` or `.atex`

**Reply by jonaro00** [reply]
Nice, `\re` seems like a good tag.

However, I just realized our current approach would only give support for alternate sequences of 1 bar length. If longer alternative sequences are to be supported, open/close tags would probably be required. Another thing to consider adding is alternate bars that are not necessarily endings (see "Other variation in repeated sections" below the section in my link from the first post), but I haven't checked if alphaTab supports those. These 2 points makes me think tags like `\rao` and `\rac` (repeat alternative open/close) are needed.

Another thing to specify would be what happens when repeat numbers are missing, such as 3 in your example. Expanding the first one so that `... | \re 1 ... | \re 4 \rc 4 ... |` would equal `... | \re (1 2 3) ... | \re 4 \rc 4 ... |` is probably reasonable (but what if the first one is missing? :wink:).

Trying to implement this seems like a fun challenge for me (if I can find time for it).

Fun that you suggest `.atx`, that is what I was using before asking. My current approach when using alphaTab is to have an alphaTex piece written in a file and let a webserver render an alphaTab html template upon requesting the `.atx` file.

**Reply by Danielku15** [reply]
> However, I just realized our current approach would only give support for alternate sequences of 1 bar length. 

You can just specify the same alternate endings on multiple bars. Each `MasterBar` holds a bitflag describing on which alternate endings it should be played. Then during midi generation and rendering these flags are respected accordingly and the endings are handled. Of course this requires the author to take care quite a bit to not write combinations which are not possible or correct according to music notation. 

```
\ro 3.3.1 | \re 1 3.2.2*2 | \re 1 4.2.2*2 | \re (2 4) \rc 4 3.3.4*4 |
```

The logic for handling repeats on playback is covered here:
https://github.com/CoderLine/alphaTab/blob/ee308e1197faf67216bedaead607cc6f499abf97/src/midi/MidiPlaybackController.ts#L41-L48 

It is fairly simple: in case a bar has flagged alternate endings, it is only played if no repeat "section" is started or if the current repeat counter allows it to be played. 

And the rendering, actually could use quite some improvement to handle correctly starts and ends of angles. I haven't invested there much😅

**Reply by jonaro00** [reply]
Very nice! Then this feature should be fairly easy to implement as you say 😄.

**Reply by jonaro00** [reply]
Hi again,

I finally gave this a try. Here is my commit: https://github.com/jonaro00/alphaTab/commit/b4028a245926ed23bba27c4c75fac368fb6d8c8e

The main functionality of the feature seems to work. There are a few things on my mind that could be discussed before I write a test and Pull Request:

- Naming inconsistencies: Alternate endings (model) vs Repeat endings (suggested tag name \re). Do we keep it like this or use the same name everywhere.
- The following tex shows two unwanted(?) behaviors that might need some fixing:
```
// Multiple single \re tags (not recommended but works)
// Slide to outside of repeat gives incorrect slide target during 2nd repeat
\ro 1.1.1 | 4.3.4*4 | \re 1 \re 3 1.1.1 | \re 2 \re 4 \rc 4 2.1{ss} | 5.1 | r

// \re tags with parentheses working as intended.
// Alternate beginnings and endings... weird behavior
\ro \re 1 1.1.1 | \re 2 2.1 | \re 3 3.1 | 4.3.4*4 | \re (1 3) 1.1.1 | \re 2 \rc 3 2.1 | r

// Getting out of control: Alternate beginnings, middles and endings
\ro \re 1 1.1.1 | \re 2 2.1 | \re 3 3.1 | 4.3.4*4 | \re 1 1.1.1 | \re 2 2.1 | \re 3 3.1 | 4.3.4*4 | \re (1 3) 1.1.1 | \re 2 \rc 3 2.1 | r
```
The first example is pretty unrelated to this feature.
The second and third example skips over the middle bars during the first and second repeats. Not sure what causes this. Is this intended (are alternate beginnings even a thing)? Is the issue in my code or elsewhere?

**Reply by Danielku15** [reply]
* Naming inconsistencies: It would be indeed better to talk about "alternate endings" and use a tag like `\ae`. As we already have `\ro` for "Repeat Open" and `\rc` for "Repeat Close", it would be confusing where the difference between close and ending is. 
* I would recommend adding tests for those examples so we can continue on this base. For the issues with the wrong repeats you might want to take this test as a reference: https://github.com/CoderLine/alphaTab/blob/478679f26440b5b80dc6125325330922f3b65acd/test/audio/MidiPlaybackController.test.ts#L58 It would allow you/me to debug very scoped why maybe an expected repeat order is not generated. 

Alternate Endings can be very tricky and there might be combinations possible that don't result in any meaningful playback order. The commit already looks fine from the general extension of the parsing. 👍 

Plus: what I like to do in case of doubts on the right behavior is to create the same file in Guitar Pro and double check the behavior and data model which is parsed. Sometimes there are small details on the data model consistency which are easy to miss unless you have the full overview.

**Reply by jonaro00** [reply]
I corrected the naming and added two tests for the `\ae` tag.

For the PlaybackController, I quickly copied the mentioned test case to investigate the weird behavior (bars getting skipped) when alternate "endings" are not at the end of a repeat. I would appreciate if you verify that the behavior expected by my test is indeed the expected behavior when using `\ae` in such ways. The current fail is: 
![image](https://user-images.githubusercontent.com/54029719/158062108-c5e44ffb-a8f7-4396-a884-518228460257.png)
...which is weird, since the test above passes just fine.

I can't think of any combination of alternate endings that would be "unplayable" yet. This sequence
```
\ro 1.1.1 | \ae 1 2.1 | 3.1 | \ae 2 4.1 | \rc 2 5.1
```
is of course weird to have written, but in my opinion it should be playable in the order (by fret) `1 2 3 5 1 3 4 5`, but currently it plays `1 2 3 1 4 5`...

The commits: https://github.com/jonaro00/alphaTab/commits/develop
Tell me if I should Pull Request at this state.

**Reply by Danielku15** [reply]
Best would be if you create already a _draft_ Pull Requests. This would allow me to have a closer look.

**Comment by jonaro00** [comment]
Another related feature request that can also be used to help show & verify the repeat functionalities is "Unfolding" repeats (can be seen in the lilypond example previously linked). This would probably be best suited as an `api.settings` property, right?

**Reply by Danielku15** [reply]
Indeed this would be also a new feature which is likely controlled through a new `api.settings` property. Hard to say at this point how to realize it best but it is likely quite a complex feature despite it's simple result. The key point is that suddenly individual model elements like notes, are rendered multiple times. alphaTab expects elements to be unique on some levels and this unfolding would break the current expectations. Relations like next/previous beats (like for ties/slurs) which are kept within the data model are broken/changed during rendering. 

One way to solve this could be to pre-process the input `Score` and create a new `Score` which has repeats unfolded within the data model itself like it would have been written without repeats. This keeps the rendering logic as it is and only introduces this preprocessing step.

**Reply by jonaro00** [reply]
Sounds good, I would have guessed the solution would look something like that.
